src/timer.c

```diff
@@ -1,155 +1,95 @@
- * Copyright (c) 2019 by magnum
+ * This file is Copyright (c) 2021 by magnum
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted.
- *
- * Portable hi-res timer.  Was a nice C++ class. Downgraded to C for john.
  */
 
 #include <stdio.h>
 #include <stdint.h>
 
 #include "timer.h"
 
-#if __MACH__
-static uint64_t sm_timebase;
-#endif
-static int sm_fGotHRTicksPerSec = 0;   // Set if we have got the above
-
-uint64_t sm_HRTicksPerSec;  // HR ticks per second (claimed)
-uint64_t sm_hrPrecision;    // HR ticks per second (observed, guess)
-uint64_t sm_cPrecision;     // clocks (ticks) per second (observed, guess)
+#if defined (_MSC_VER) || defined (__MINGW32__) || defined (__CYGWIN32__)
 
-void sTimer_Init(sTimer *t)
+/*
+ * QueryPerformanceCounter(LARGE_INTEGER*) gives "counts" (ticks)
+ * QueryPerformanceFrequency(LARGE_INTEGER*) gives ticks/s
+ */
+uint64_t john_get_nano(void)
 {
-	t->m_fRunning = 0;
-	t->m_cStartTime = 0;
-	t->m_cEndTime = 0;
-	t->m_dAccumSeconds = 0;
-	HRZERO(t->m_hrStartTime);
-	HRZERO(t->m_hrEndTime);
-	if (!sm_fGotHRTicksPerSec) {
-		// What's the lowest digit set non-zero in a clock() call
-		// That's a fair indication what the precision is likely to be.
-		// Note - this isn't actually used
-		int i;
-
-		sm_fGotHRTicksPerSec = 1;
-		sm_cPrecision = 0;
-		for (i = 0; i < 10; ++i) {
-			clock_t heuristicTimeTest = clock();
-
-			if (heuristicTimeTest % 10) {
-				sm_cPrecision = CLOCKS_PER_SEC;
-				break;
-			}
-			else if (heuristicTimeTest % 100) {
-				if (!sm_cPrecision || sm_cPrecision < CLOCKS_PER_SEC / 10)
-					sm_cPrecision = CLOCKS_PER_SEC / 10;
-			}
-			else if (heuristicTimeTest % 1000) {
-				if (!sm_cPrecision || sm_cPrecision < CLOCKS_PER_SEC / 100)
-					sm_cPrecision = CLOCKS_PER_SEC / 100;
-			}
-			else if (heuristicTimeTest % 10000) {
-				if (!sm_cPrecision || sm_cPrecision > CLOCKS_PER_SEC / 1000)
-					sm_cPrecision = CLOCKS_PER_SEC / 1000;
-			}
-			else {
-				if (!sm_cPrecision || sm_cPrecision > CLOCKS_PER_SEC / 10000)
-					sm_cPrecision = CLOCKS_PER_SEC / 10000;
-			}
-		}
-
-		// Find the claimed resolution of the high res timer
-		// Then find the most likely real rate by waiting for it to change.
-		// Note - I've frequently seen missed beats, and therefore a
-		// 0.000001 reality gets reported as a 0.000002.
-		// Note - this also isn't actually used, all that matters is
-		// whether HRTicksPerSec has a non-zero value or not.
-		HRGETTICKS_PER_SEC(sm_HRTicksPerSec);
-
-		if (sm_HRTicksPerSec) {
-			hr_timer start, end;
-
-			HRSETCURRENT(start);
-			do {
-				HRSETCURRENT(end);
-			} while (HRGETTICKS(end) == HRGETTICKS(start));
-
-			sm_hrPrecision = sm_HRTicksPerSec / (HRGETTICKS(end) -
-			                  HRGETTICKS(start));
-		}
+	static uint64_t ticks_per_sec;
+	LARGE_INTEGER t;
+
+	if (!ticks_per_sec) {
+		QueryPerformanceFrequency(&t);
+		ticks_per_sec = t.QuadPart;
 	}
+
+	QueryPerformanceCounter(&t);
```

Comment:
This appears to rely on `timer.h` having included `<windows.h>`, and similar for MACH below. I think it's better to move those `#include` directives from `timer.h` to `timer.c` now that we no longer use those things in macros.